### PR TITLE
BUG: DataFrame.__setitem__ sometimes operating inplace

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

The fixed bug is the currently-xfailed test test_setitem_same_dtype_not_inplace (see #39510, which has its own subsection in the 1.3.0 release notes).  Pretty much all the other altered tests are changed because those tests rely on the incorrect behavior.

The main non-test change is the addition of an `inplace` check in `BlockManager.iset`.  We want inplace to be False when this is reached via `DataFrame.__setitem__` and True when reached via `loc/iloc.__setitem__`

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -1217,7 +1217,7 @@ def _check_is_chained_assignment_possible(self) -> bool:
         return super()._check_is_chained_assignment_possible()
 
     def _maybe_update_cacher(
-        self, clear: bool = False, verify_is_copy: bool = True
+        self, clear: bool = False, verify_is_copy: bool = True, inplace: bool = False
```
Comment (jreback) [inline pandas/core/series.py:1220]: why is this different than the superclass? this is really strange

**Review by jreback — CHANGES_REQUESTED** [review]
can you make a user note of *when* this is changed, its not obvious at all.

**Comment by jreback** [comment]
cc @jorisvandenbossche

**Comment by jbrockmendel** [comment]
cc @phofl

File: pandas/core/series.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/series.py:1220]: i dont understand the question.  the signature is the same here and in the superclass

Comment (jreback) [reply pandas/core/series.py:1225]: what i mean is why are these inplace args different in the super and sub classes?

Comment (jbrockmendel) [reply pandas/core/series.py:1225]: > why are these inplace args different

they aren't. Here Series has `inplace: bool = False`, up in generic NDFrame has `inplace: bool_t = False`

Comment (jreback) [reply pandas/core/series.py:1225]: oh maybe i was reading something incorrectly

**Comment by jbrockmendel** [comment]
@jorisvandenbossche @phofl gentle ping

**Review by phofl — COMMENTED** [review]
This lgtm in general. I think we should add tests to cover this:

```
@pytest.mark.parametrize("indexer", ["a", ["a"], [True, False]])
@pytest.mark.parametrize(
    "value, set_value",
    [
        (1, 5),
        (1.0, 5.0),
        (Timestamp("2020-12-31"), Timestamp("2021-12-31")),
        ("a", "b"),
    ],
)
def test_setitem_not_operating_inplace(value, set_value, indexer):
    df = DataFrame({"a": value}, index=[0, 1])
    expected = df.copy()
    view = df[:]
    df[indexer] = set_value
    tm.assert_frame_equal(view, expected)
```

It looks like as the boolean indexer case is still operating inplace

**Comment by jbrockmendel** [comment]
@jreback gentle ping, doesn't sound like the others are weighing in

**Comment by phofl** [comment]
I think this looks good. Have to figure out the boolean case in the future, but otherwise is good

**Comment by jreback** [comment]
let me look

**Comment by jbrockmendel** [comment]
gentle ping; has a bearing on #40380

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
cool, ideally can simplify the caching checks at some point (e.g. the maybe_update_cacher is getting more complicated).

**Comment by SimeonStoykovQC** [comment]
`df.assign` uses `DataFrame.__setitem__`:

https://github.com/pandas-dev/pandas/blob/965ceca9fd796940050d6fc817707bba1c4f9bff/pandas/core/frame.py#L4841-L4845

> We want inplace to be False when this is reached via `DataFrame.__setitem__` and True when reached via `loc/iloc.__setitem__`

Did this change intend to modify `df.assign`'s behavior? Or do you think `assign` should be using the in-place mode again through `loc/iloc`?

**Comment by jonashaag** [comment]
@SimeonStoykovQC no, `assign` should not modify the existing data so not be in-place

**Comment by jonashaag** [comment]
@SimeonStoykovQC I misunderstood what you meant. You meant to use `[...]` instead of `.loc[:, ...]` on the copy, right? That makes sense if `.loc` is faster (which I don't know)

**Comment by phofl** [comment]
No, the assign change is intended.

**Comment by jonashaag** [comment]
It caused a 10x performance regression in some of our code. We will try to come up with a reproducer.

**Comment by SimeonStoykovQC** [comment]
> No, the assign change is intended.

In the meantime, could you elaborate? What's wrong with in-place allocations if [`df.assign` is creating a deep copy (prior version 2.0)](https://github.com/pandas-dev/pandas/blob/2e218d10984e9919f0296931d92ea851c6a6faf5/pandas/core/frame.py#L4886)?

**Comment by phofl** [comment]
Yeah I know, but this was fixed for 2.0, we had a nasty performance bottleneck in 1.5 unfortunately. Everything is back to normal with 2.0

**Pull request merged**